libs/execution_context/ebpf_maps.c

```diff
@@ -866,82 +878,77 @@ _update_entry_per_cpu(
     return EBPF_SUCCESS;
 }
 
-// Given two keys, compute the number of bits that match.
-static size_t
-_lpm_trie_matching_key_prefix_length(
-    _In_reads_((key_length_a + 7) / 8) const uint8_t* key_a,
-    size_t key_length_a,
-    _In_reads_((key_length_b + 7) / 8) const uint8_t* key_b,
-    size_t key_length_b)
+static void
+_lpm_extract(_In_ const uint8_t* value, _Outptr_ const uint8_t** data, _Out_ size_t* length_in_bits)
 {
-    size_t maximum_matching_length = key_length_a > key_length_b ? key_length_b : key_length_a;
-    size_t remaining_key_length_in_bits = maximum_matching_length;
-    uint8_t key_a_last_byte;
-    uint8_t key_b_last_byte;
-    size_t index;
-
-    while (remaining_key_length_in_bits >= 8) {
-        index = (maximum_matching_length - remaining_key_length_in_bits) / 8;
-        if (key_a[index] != key_b[index]) {
-            break;
-        }
-        remaining_key_length_in_bits -= 8;
-    }
-    if (remaining_key_length_in_bits == 0) {
-        return maximum_matching_length;
-    }
-
-    index = (maximum_matching_length - remaining_key_length_in_bits) / 8;
-    key_a_last_byte = key_a[index];
-    key_b_last_byte = key_b[index];
+    uint32_t prefix_length = *(uint32_t*)value;
+    *data = value;
+    *length_in_bits = sizeof(uint32_t) * 8 + prefix_length;
+}
 
-    while (remaining_key_length_in_bits) {
-        if ((key_a_last_byte & 0x80) != (key_b_last_byte & 0x80)) {
-            break;
-        }
-        remaining_key_length_in_bits--;
-        key_a_last_byte <<= 1;
-        key_b_last_byte <<= 1;
-    }
-    return maximum_matching_length - remaining_key_length_in_bits;
+static ebpf_core_map_t*
+_create_lpm_map(_In_ const ebpf_map_definition_in_memory_t* map_definition)
+{
+    ebpf_core_lpm_map_t* map = (ebpf_core_lpm_map_t*)_create_hash_map_internal(
+        EBPF_OFFSET_OF(ebpf_core_lpm_map_t, prefix_in_use) + map_definition->key_size, map_definition, _lpm_extract);
+    map->max_prefix = (map_definition->key_size - sizeof(uint32_t)) * 8;
```

Comment:
BUG: check for map == null before dereferencing map.  otherwise out-of-memory will result in a bugcheck.